### PR TITLE
feat(purl): add upstream to qualifiers

### LINKED ISSUE #3942: referenceLocator purl does not include upstream information about system packages
## Description

The `purl` information does not include the `upstream` package that a specific OS package is coming from. For instance:

`$ trivy image --format spdx-json debian:latest > debian.json`

The package `libssl1.1` includes this information `"referenceLocator": "pkg:deb/debian/libssl1.1@1.1.1n-0+deb11u4?distro=debian-11.6"`. The issue of not including the `upstream` information from the package into the `purl` is there is no way to detect CVEs based on that information as those ones are linked to the upstream package.

The correct information should be `"referenceLocator": "pkg:deb/debian/libssl1.1@1.1.1n-0+deb11u4?upstream=openssl&distro=debian-11.6"`.

## What did you expect to happen?

Include the `upstream` information from the OS packages metadata as part of the sBOM. It does not appear into the SPDX or CycloneDX `purl`.

## What happened instead?

No `upstream` information as part of the `purl`.

## Output of `trivy -v`:

```
Version: 0.38.3
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-03-31 06:17:32.691408303 +0000 UTC
  NextUpdate: 2023-03-31 12:17:32.691407903 +0000 UTC
  DownloadedAt: 2023-03-31 06:53:06.842877 +0000 UTC
```

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
Add `upstream`(SrcName) field to purl `qualifiers`.

Before:
```
pkg:rpm/centos/openssl-libs@1.0.2k-16.el7?arch=x86_64&epoch=1&distro=centos-7.6.1810
```

After:
```
pkg:rpm/centos/openssl-libs@1.0.2k-16.el7?arch=x86_64&epoch=1&upstream=openssl&distro=centos-7.6.1810
```

## Related issues
- Close #3942

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/sbom/cyclonedx/marshal_test.go [hunk 1]
```diff
@@ -224,14 +224,14 @@ func TestMarshaler_Marshal(t *testing.T) {
 				},
 				Components: &[]cdx.Component{
 					{
-						BOMRef:  "pkg:rpm/centos/binutils@2.30-93.el8?arch=aarch64&distro=centos-8.3.2011",
+						BOMRef:  "pkg:rpm/centos/binutils@2.30-93.el8?arch=aarch64&upstream=binutils&distro=centos-8.3.2011",
```
Comment (beltran-rubo) [inline pkg/sbom/cyclonedx/marshal_test.go:227]: The `upstream` reference works for DPKG packages.
For RPMs the `upstream` should point to the SRC package that is coming from e.g. `upstream=vim-8.2.2637-20.el9_1.src.rpm`

Comment (DmitriyLewen) [reply pkg/sbom/cyclonedx/marshal_test.go:227]: Thanks for information. I will try to fix it.

Comment (DmitriyLewen) [reply pkg/sbom/cyclonedx/marshal_test.go]: @beltran-rubo i updated logic.
Do you have time to check it out?

Comment (beltran-rubo) [reply pkg/sbom/cyclonedx/marshal_test.go]: It looks correct to me. For `apk` packages the {{purl}} includes the {{upstream}} with the name of the package that comes from only, same approach than Deb packages.

Comment (DmitriyLewen) [reply pkg/sbom/cyclonedx/marshal_test.go]: Cool! Thanks for your help!

**Comment by knqyf263** [comment]
We don't plan to do that at the moment. It might change according to the spec update.

**Pull request closed without merging**